### PR TITLE
[TE] expand future time series endpoint to support time-range aggregation

### PR DESCRIPTION
**Pull request by apucher**
Enhanced prototype for a metric time series endpoint that supports online transformation and aggregation across multiple time ranges. This is a necessary building block for complex baselines in TE's frontend, such as multi-week averages. Transformations can still be chained, aggregations return separate result blocks. The PR furthermore adds several transformations and performs refactoring to add a battery of unit tests.

Supported time range aggregation: SUM, PRODUCT, MEAN, MEDIAN, STD, MIN, MAX

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/dashboard/resource/TimeSeriesResourceTest.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    Map<String, List<? extends Number>> ts = out.get(RANGE);
+
+    Assert.assertEquals(ts.get(COL_TIME), Arrays.asList(0L, 10L, 20L, 30L, 40L, 50L));
+    assertEquals(ts.get("0"), Arrays.asList(1d, 1d, 0d, 0d, 1d, 1d));
+  }
+
+  @Test
+  public void testTranformationForwardFill() throws Exception {
+    Map<String, Map<String, List<? extends Number>>> out = this.resource.getTimeSeries(
+        "0", RANGE, null, null, "forwardfill", null);
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/dashboard/resource/TimeSeriesResourceTest.java]: fillforward?

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/dashboard/resource/TimeSeriesResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    Assert.assertEquals(ts.get(COL_TIME), Arrays.asList(0L, 1L, 2L, 3L, 4L, 5L));
+    assertEquals(ts.get("0"), Arrays.asList(1d, 0d, 1d, 0d, 1d, null));
+    assertEquals(ts.get("1"), Arrays.asList(4d, 1d, 0d, 1d, 4d, null));
+
+  }
+
+  @Test
+  public void testMultiTransformationMultiAggregationMultiRange() throws Exception {
+    Map<String, Map<String, List<? extends Number>>> out = this.resource.getTimeSeries(
+        "0,1", RANGE + ",20:80", null, null, "forwardfill,cumulative", "sum,product");
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/dashboard/resource/TimeSeriesResourceTest.java]: fillforward?

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/dashboard/resource/TimeSeriesResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+  @Test(expectedExceptions = IllegalArgumentException.class)
+  public void testTranformationChangeNoGranularityFail() throws Exception {
+    this.resource.getTimeSeries("0", RANGE, null, null, "timestamp", null);
+  }
+
+  @Test
+  public void testMultiTranformation() throws Exception {
+    Map<String, Map<String, List<? extends Number>>> out = this.resource.getTimeSeries(
+        "2", RANGE, null, null, "change,forwardfill,cumulative", null);
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/dashboard/resource/TimeSeriesResourceTest.java]: fillforward?

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/dashboard/resource/TimeSeriesResourceTest.java (hunk 1 above)
Comment (apucher) [reply thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/dashboard/resource/TimeSeriesResourceTest.java]: thanks. good catch. travis complained too.

**Pull request merged**